### PR TITLE
feat: generics (Record<K,V>, user-defined generic aliases, Array<T>/Schema<T>)

### PR DESCRIPTION
**Pull request by egonSchiele**
## Summary

Adds generics to the Agency type system, per the design at [docs/superpowers/specs/2026-05-18-generics-design.md](file:///Users/adityabhargava/agency-lang/packages/agency-lang/docs/superpowers/specs/2026-05-18-generics-design.md) and the plan at [docs/superpowers/plans/2026-05-19-generics-design.md](file:///Users/adityabhargava/agency-lang/packages/agency-lang/docs/superpowers/plans/2026-05-19-generics-design.md).

User-visible additions:

- **`Record<K, V>`** as a built-in generic with zod (`z.record(K, V)`) and TS (`Record<K, V>`) codegen. Treated as **covariant** in both `K` and `V` for ergonomics (the variance gotcha is documented in `docs/site/guide/types.md`).
- **User-defined generic type aliases**: `type Container<T> = { value: T }`, with defaults (`type StringMap<V = any> = ...`), and full recursion support (`type Tree<T> = { value: T, children: Tree<T>[] }`).
- **`Array<T>` and `Schema<T>`** are now built-in generic forms, equivalent to `T[]` and the type of `schema(T)`.
- **For-loops over Records**: `for (k in record)` iterates the keys.
- The `object` primitive still works; `Record<string, any>` is now the preferred replacement (removal of `object` is deferred per the plan — Task 15 in the plan is independent).

## Architecture

- New AST nodes: `GenericType`, `TypeParam`, `TypeAliasEntry`. The parser is policy-free: every `Name<...>` becomes a `genericType`.
- The type-alias registry (`ScopedTypeAliases`) is widened to carry `typeParams` through cross-module imports.
- A new declarative tree transformer `mapTypes` (sibling to `visitTypes`) encapsulates the per-variant recursion. `substituteTypeParams` is built on top of it in two lines.
- `resolveType` was split into a public single-arg API and a private `resolveTypeWithGuard` that threads a recursion-tracking set. Built-in generics (`Array`, `Schema`, `Record`) and user-defined aliases are handled in the same place.

## Validation

- 4001 tests pass (was 3904 before this branch — 97 added).
- `pnpm run lint:structure` and `tsc --noEmit` both clean.
- Cross-module test confirms `typeParams` propagate through imports.
- End-to-end smoke: `tests/typescriptGenerator/recordIndex.agency` exercises declaration, index write, index read, and for-in iteration over a `Record<string, string>` through codegen and runtime.

## Notes on deferred work

- **Task 15 (remove `object`)**: deferred. The plan explicitly calls this out as independent. It touches many agencyGenerator tests and class-method codegen and is a separate cleanup.
- **Task 17 (full edge-case fixture sweep)**: most of the cases are already covered by focused unit tests under `lib/typeChecker/`. A consolidated fixture suite can be a follow-up.

### CODE REVIEW

File: packages/agency-lang/docs/site/guide/types.md [hunk 1]
```diff
@@ -18,7 +18,7 @@ Primitive types:
 - `boolean`
 - `null`
 - `undefined` (treated as null)
-- `object` (equivalent to `Record<string, any>`)
+- `object` (equivalent to `Record<string, any>` — prefer `Record<string, any>` directly in new code)
```
Comment (egonSchiele) [inline packages/agency-lang/docs/site/guide/types.md]: change to remove all mentions of `object`

File: packages/agency-lang/docs/site/guide/types.md [hunk 2]
```diff
@@ -57,6 +57,78 @@ type User = {
...
+  print("{k}: {votes[k]}")
+}
+```
+
+`Record<K, V>` is a built-in generic with two type arguments: the key type and the value type. The key type must be `string`, `number`, a string/number literal, or a union of those.
+
+You can also use a literal-key union for `K` to require a closed set of keys:
+
+```ts
+let status: Record<"active" | "inactive", number> = {
```
Comment (egonSchiele) [inline packages/agency-lang/docs/site/guide/types.md:77]: what if they use a type that is a union for the key?

File: packages/agency-lang/lib/backends/typescriptGenerator/typeToZodSchema.test.ts [hunk 3]
```diff
@@ -107,3 +107,51 @@ describe("mapTypeToValidationSchema", () => {
...
+          { type: "primitiveType", value: "string" },
+          { type: "primitiveType", value: "number" },
+        ],
+      },
+      {},
+    );
+    expect(result).toBe("z.record(z.string(), z.number())");
+  });
+
+  it("emits z.record nested inside z.array", () => {
```
Comment (egonSchiele) [inline packages/agency-lang/lib/backends/typescriptGenerator/typeToZodSchema.test.ts:127]: what about nested records?

File: packages/agency-lang/lib/backends/agencyGenerator.ts [hunk 4]
```diff
@@ -464,10 +464,19 @@ export class AgencyGenerator {
     this.typeAliases[node.aliasName] = node.aliasedType;
     const aliasedTypeStr = this.aliasedTypeToString(node.aliasedType);
     const exportPrefix = node.exported ? "export " : "";
+    const typeParamsStr = node.typeParams?.length
```
Comment (egonSchiele) [inline packages/agency-lang/lib/backends/agencyGenerator.ts]: These nested terneries are hard to read.

File: packages/agency-lang/lib/backends/typescriptBuilder/scopeManager.ts [hunk 5]
```diff
@@ -63,6 +63,23 @@ export class ScopeManager {
...
+    }
+    return out;
+  }
+
+  /**
+   * Full TypeAliasEntry-form alias registry for callers (resolveTypeDeep)
+   * that need type-parameter metadata to substitute user-defined generic
+   * aliases.
+   */
+  visibleTypeAliasesFull() {
```
Comment (egonSchiele) [inline packages/agency-lang/lib/backends/typescriptBuilder/scopeManager.ts]: add a return type

File: packages/agency-lang/lib/runtime/runner.ts [hunk 6]
```diff
@@ -403,7 +403,14 @@ export class Runner {
 
     this.frame.locals[iterKey] = this.frame.locals[iterKey] ?? 0;
 
-    for (let i = 0; i < items.length; i++) {
+    // Records (plain objects) iterate by key. Arrays iterate by element.
+    const iterable: any[] = Array.isArray(items)
```
Comment (egonSchiele) [inline packages/agency-lang/lib/runtime/runner.ts]: These nested ternaries are hard to read.

File: packages/agency-lang/lib/typeChecker/assignability.ts [hunk 7]
```diff
@@ -1,18 +1,253 @@
...
+        entry.body,
+        entry.typeParams.map((p) => p.name),
+        args,
+      );
+      return resolveTypeWithGuard(substituted, typeAliases, next);
+    }
+    return resolveTypeWithGuard(entry.body, typeAliases, next);
+  }
+
+  if (vt.type === "genericType") {
```
Comment (egonSchiele) [inline packages/agency-lang/lib/typeChecker/assignability.ts:82]: prefer separating the data and implementation here. So add a constant that specifies the shape of these built-in types, like

```
[{ typeName: "Array", typeArgs: 1}]
```

File: packages/agency-lang/lib/typeChecker/assignability.ts [hunk 8]
```diff
@@ -1,18 +1,253 @@
...
+ * defined generic aliases.
+ */
+export function resolveTypeDeep(
+  t: VariableType,
+  typeAliases: Record<string, TypeAliasEntry>,
+): VariableType {
+  let current = t;
+  // Iterate until a pass produces no change. Generic-alias depth is bounded
+  // in practice; the cap exists only to prevent runaway loops on a bug.
+  for (let i = 0; i < 32; i++) {
```
Comment (egonSchiele) [inline packages/agency-lang/lib/typeChecker/assignability.ts]: magic number

File: packages/agency-lang/lib/typeChecker/genericCrossModule.test.ts [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+  const symbolTable = SymbolTable.build(absPath, {});
+  const parseResult = parseAgency(source, {});
+  if (!parseResult.success)
+    throw new Error("Parse failed: " + parseResult.message);
+  const program = parseResult.result;
+  const info = buildCompilationUnit(program, symbolTable, absPath, source);
+  return typeCheck(program, {}, info).errors;
+}
+
+describe("cross-module generic type alias", () => {
```
Comment (egonSchiele) [inline packages/agency-lang/lib/typeChecker/genericCrossModule.test.ts]: would these be better written as agency tests? You could import a type and use it as a schema for validation to test

File: packages/agency-lang/CHANGELOG.md [hunk 10]
```diff
@@ -7,6 +7,11 @@
 - Const reassignment checks in typechecker
 - `parseJSON` built-in function
 - Fix string interpolation parser for concatenating function calls and strings
+- **Generics**: built-in `Record<K, V>` type with `z.record` codegen and TS `Record<K, V>` output; covariant for ergonomics (documented variance gotcha in [docs/site/guide/types.md](file:///Users/adityabhargava/agency-lang/packages/agency-lang/docs/site/guide/types.md))
```
Comment (egonSchiele) [inline packages/agency-lang/CHANGELOG.md]: don't update the changelog
